File: chapters/chapter2/scripts/matlab/getRefinedValues.m
function [x,y,dy]=getRefinedValues(u,xl,d,k)
  d_ref=linspace(0,1,k+1); % refined basis element
  [E,D]=interpolatingMat(d,d_ref);
  n=numel(xl); m=numel(d)-1;
  x_ref=(n-1)*k+1; % number of all refined mesh points
  x=zeros(x_ref,1); y=zeros(size(x)); dy=zeros(size(x));
  % Iterate over finite elements and refine them 
  for l=1:n-1
    h=xl(l+1)-xl(l); % size of element l
    i_c=(l-1)*m+(1:m+1); % coarse point indices
    i_ref=(l-1)*k+(1:k+1); % refined points indices
    x(i_ref)=linspace(xl(l),xl(l+1),k+1)';
    y(i_ref)=E*u(i_c);
    dy(i_ref)=D*u(i_c)/h;
  end
end

File: chapters/chapter2/scripts/python/getRefinedValues.py
def getRefinedValues(u,xl,d,k):
    d_ref=numpy.linspace(0,1,k+1) # refined basis element
    E,D=interpolatingMat(d,d_ref)
    n=numpy.size(xl); m=numpy.size(d)-1
    x_ref=(n-1)*k+1; # number of all refined mesh points
    x=numpy.zeros((x_ref,))
    y=numpy.zeros(x.shape)
    dy=numpy.zeros(x.shape)
    for l in range(n-1):
        h=xl[l+1]-xl[l]  #size of element l
        i_c=l*m+numpy.arange(m+1) # coarse point indices
        i_ref=l*k+numpy.arange(k+1)  # refined points indices
        x[i_ref]=numpy.linspace(xl[l],xl[l+1],k+1)
        y[i_ref]=E @ u[i_c] 
        dy[i_ref]=D @ u[i_c]/h 
    return x,y,dy
